File: app/src/main/java/me/cargoapp/cargo/messaging/MessagingNotificationInteracter.java
package me.cargoapp.cargo.messaging;

import android.app.PendingIntent;
import android.app.RemoteInput;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;

import org.greenrobot.eventbus.EventBus;

import me.cargoapp.cargo.event.message.DismissMessageNotificationAction;

public class MessagingNotificationInteracter {
    static public boolean reply(Context context, MessagingNotificationParser.NotificationParserResult result, String text) {
        if (result.application == MessagingApplication.NONE) return false;
        if (TextUtils.isEmpty(text.trim())) return false;

        int actionIndex;
        int remoteInputIndex;
        switch (result.application) {
            case SMS:
                actionIndex = 0;
                remoteInputIndex = 0;
                break;
            case MESSENGER:
                actionIndex = 1;
                remoteInputIndex = 0;
                break;
            default:
                return false;
        }

        final RemoteInput[] remoteInputs = result.sbn.getNotification().actions[actionIndex].getRemoteInputs();
        Intent intent = new Intent();
        Bundle results = new Bundle();
        results.putString(remoteInputs[remoteInputIndex].getResultKey(), text.trim());
        RemoteInput.addResultsToIntent(remoteInputs, intent, results);

        try {
            result.sbn.getNotification().actions[actionIndex].actionIntent.send(context, 0, intent);
            dismiss(result);
            return true;
        } catch (PendingIntent.CanceledException e) {
            e.printStackTrace();
            return false;
        }
    }

    static public void dismiss(MessagingNotificationParser.NotificationParserResult result) {
        EventBus.getDefault().post(new DismissMessageNotificationAction(result));
    }
}

File: app/src/main/java/me/cargoapp/cargo/messaging/MessagingNotificationInteracter.kt
package me.cargoapp.cargo.messaging

import android.app.PendingIntent
import android.app.RemoteInput
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils

import org.greenrobot.eventbus.EventBus

import me.cargoapp.cargo.event.message.DismissMessageNotificationAction

object MessagingNotificationInteracter {
    fun reply(context: Context, result: MessagingNotificationParser.NotificationParserResult, text: String): Boolean {
        if (result.application === MessagingApplication.NONE) return false
        if (TextUtils.isEmpty(text.trim { it <= ' ' })) return false

        val actionIndex: Int
        val remoteInputIndex: Int
        when (result.application) {
            MessagingApplication.SMS -> {
                actionIndex = 0
                remoteInputIndex = 0
            }
            MessagingApplication.MESSENGER -> {
                actionIndex = 1
                remoteInputIndex = 0
            }
            else -> return false
        }

        val remoteInputs = result.sbn.notification.actions[actionIndex].remoteInputs
        val intent = Intent()
        val results = Bundle()
        results.putString(remoteInputs[remoteInputIndex].resultKey, text.trim { it <= ' ' })
        RemoteInput.addResultsToIntent(remoteInputs, intent, results)

        try {
            result.sbn.notification.actions[actionIndex].actionIntent.send(context, 0, intent)
            dismiss(result)
            return true
        } catch (e: PendingIntent.CanceledException) {
            e.printStackTrace()
            return false
        }

    }

    fun dismiss(result: MessagingNotificationParser.NotificationParserResult) {
        EventBus.getDefault().post(DismissMessageNotificationAction(result))
    }
}
